test/server_test.go

```diff
@@ -0,0 +1,284 @@
+/*
+ *
+ * Copyright 2014 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package test
+
+import (
+	"context"
+	"io"
+	"testing"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	testpb "google.golang.org/grpc/test/grpc_testing"
+)
+
+type ctxKey string
+
+func (s) TestChainUnaryServerInterceptor(t *testing.T) {
+	var (
+		firstIntKey  = ctxKey("firstIntKey")
+		secondIntKey = ctxKey("secondIntKey")
+	)
+
+	firstInt := func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		if ctx.Value(firstIntKey) != nil {
+			t.Fatalf("first interceptor should not have %v in context", firstIntKey)
```

Comment:
Please make `fatal`s return an error (`status.Errorf()`) instead. The client can check the returned error and fail the test.

This will be called from a service handler goroutine. Fatal should only be called in the main test goroutine.

Reply:
Surely, full gRPC server should return proper errors and let client handle them...
Thanks!